### PR TITLE
Option to invert normals

### PR DESCRIPTION
**Pull request by nickkeepfer**
Keyword argument option to invert normals if neccessary on the makemesh_GeometryBasics function.
(normals can are inverted if mesh surface is not a closed volume)

### CODE REVIEW

**Comment by t-bltg** [comment]
What about adding a `normal_sign` attribute to the `MC` struct, and multiplying the magnitude of the normals with this sign in https://github.com/JuliaGeometry/MarchingCubes.jl/blob/93f26baa707d9a7af00ff9203a3286b939326e8f/src/MarchingCubes.jl#L629 ?

**Comment by nickkeepfer** [comment]
Yep, I agree, much more succinct!

File: src/example.jl [hunk 1]
```diff
@@ -72,7 +72,7 @@ makemesh_Meshes(Meshes::Module, m::MC) = begin
     Meshes.SimpleMesh(points, [tris;])
 end
 
-makemesh_GeometryBasics(GeometryBasics::Module, m::MC) = begin
+makemesh_GeometryBasics(GeometryBasics::Module, m::MC, invert_normals::Bool=false) = begin
```
Comment (t-bltg) [inline src/example.jl:75]: I think this can be removed as well.

Comment (nickkeepfer) [reply src/example.jl:75]: Oops, good catch

File: src/MarchingCubes.jl [hunk 2]
```diff
@@ -64,7 +64,7 @@ struct MC{F,I}
     x::RefValue{Vector{F}}
     y::RefValue{Vector{F}}
     z::RefValue{Vector{F}}
-
+    normal_sign::Int
```
Comment (t-bltg) [inline src/MarchingCubes.jl:67]: Can you add a comment here, saying that `+1` corresponds to the outward pointing normal and `-1` to the inward one ?

File: src/MarchingCubes.jl [hunk 3]
```diff
@@ -629,7 +630,7 @@ add_y_vertex(m::MC{F}, vol, cb, i, j, k) where {F} = begin
         (mag = norm(m.nrm)) > eps(F) && (m.nrm ./= mag)
     end
     push!(m.vertices, Vertex(i - 1, j - 1 + u, k - 1))
-    push!(m.normals, Normal(m.nrm))
+    push!(m.normals, Normal(m.normal_sign * m.nrm))
```
Comment (t-bltg) [inline src/MarchingCubes.jl:633]: I think that computationally it would be more efficient to do the multiplication here:
https://github.com/JuliaGeometry/MarchingCubes.jl/blob/main/src/MarchingCubes.jl#L613-L615

Comment (nickkeepfer) [reply src/MarchingCubes.jl:633]: As in: (mag = m.normal_sign*norm(m.nrm)) > eps(F) && (m.nrm ./= mag), or on each component individually?

Comment (t-bltg) [reply src/MarchingCubes.jl:633]: I think (haven't checked) that `m.normal_sign * m.nrm` can allocate, but `m.normal_sign*norm(m.nrm)` is a non-allocating solution.

However, it is usually implied that the term `magnitude` is positive, so I find it confusing that it could be negative, so my preference would go to multiply each component individually ...

File: src/MarchingCubes.jl [hunk 4]
```diff
@@ -64,7 +64,7 @@ struct MC{F,I}
     x::RefValue{Vector{F}}
     y::RefValue{Vector{F}}
     z::RefValue{Vector{F}}
-
+    normal_sign::Int # direction of normal vectors (+1 for outward / -1 for inward)
```
Comment (t-bltg) [inline src/MarchingCubes.jl:67]:
```suggestion
    normal_sign::Int  # direction of normal vectors (+1 for outward / -1 for inward)
```

nitpicking on the number of spaces :)

File: src/MarchingCubes.jl (hunk 3 above)
Comment (nickkeepfer) [reply src/MarchingCubes.jl]: Not sure how you want to do it for line 690, in `add_c_vertex`, 
Do you think `m.nrm .+= m.normal_sign * m.normals[n])` for each line?

Comment (t-bltg) [reply src/MarchingCubes.jl]: Good point, I haven't thought about it.

Regarding my initial comment https://github.com/JuliaGeometry/MarchingCubes.jl/pull/14#discussion_r1088185937, I was wrong: you can use `Normal(m.normal_sign * m.nrm)` everywhere, since it is non-allocating.

Tested with:
```julia
julia> using MarchingCubes, BenchmarkTools
julia>  @btime march($mc)
  8.667 ms (0 allocations: 0 bytes)
```

Comment (nickkeepfer) [reply src/MarchingCubes.jl]: In that case, do you want me to revert them all back to `Normal(m.normal_sign * m.nrm)`?

Comment (t-bltg) [reply src/MarchingCubes.jl]: Yes, exactly.
Also since the formulation `m.normal_sign * (1 - u) * ∇x(vol, i, j, k, m.nx) + u * ∇x(vol, i + 1, j, k, m.nx)` is wrong (you'd need parentheses, because of the `+` sign).

Can you add a test with the negative sign ?

Would it be possible to post an image of the example in `Makie` using the negative sign (I think we should see the shading change ...) ?

Comment (t-bltg) [reply src/MarchingCubes.jl]: Not necessary, just to visually check here that everything is correct.

For the `test/` regression check, you can e.g. check that you obtain the same number of normals, and compute the `L2 norm` of `mc.normals` (it shouldn't change, compare with `≈` against the positive sign case).

Can you also add a `normal_sign ∈ (-1, +1) || throw(ArgumentError("`normal_sign` should be either -1 or +1"))` check in the constructor, and check that it throws correctly in the tests ?

File: src/MarchingCubes.jl [hunk 5]
```diff
@@ -71,7 +71,7 @@ struct MC{F,I}
         x::AbstractVector{F} = F[],
         y::AbstractVector{F} = F[],
         z::AbstractVector{F} = F[],
-        normal_sign::Int = Int,
+        normal_sign::Type{G} = Int,
```
Comment (t-bltg) [inline src/MarchingCubes.jl]: @nickkeepfer I see you struggle with this, I think you want `normal_sign::Int = 1,` as constructor default value.

File: src/MarchingCubes.jl [hunk 6]
```diff
@@ -86,7 +87,9 @@ struct MC{F,I}
             Ref(x),
             Ref(y),
             Ref(z),
+            1,
```
Comment (t-bltg) [inline src/MarchingCubes.jl:90]: And propagate here
```suggestion
            normal_sign,
```

File: src/MarchingCubes.jl (hunk 5 above)
Comment (nickkeepfer) [reply src/MarchingCubes.jl]: Yes, constructors are new to me, thank you very much, I was indeed struggling!

Comment (t-bltg) [reply src/MarchingCubes.jl]: That's how you learn ;)

File: test/runtests.jl [hunk 7]
```diff
@@ -118,7 +118,7 @@ end
     m2 = MC(dat,normal_sign=-1)
     march(m2)
 
-    @test norm(m1.normals,m1.normals) ≈ norm(m2.normals,m2.normals)
+    @test norm(m1.normals) ≈ norm(m2.normals)
```
Comment (t-bltg) [inline test/runtests.jl:121]: You want to check that each normal has the same magnitude (they are computed in the same order, so this should be valid):
```julia
    @test all(norm.(m1.normals) .≈ norm.(m2.normals))
```

**Comment by nickkeepfer** [comment]
I think (finally) I've managed to implement it successfully. This has been a good learning experience!

**Comment by t-bltg** [comment]
Great, thanks, I'll leave a view comments and we should be good to go.

File: src/MarchingCubes.jl [hunk 8]
```diff
@@ -616,33 +620,33 @@ add_x_vertex(m::MC{F}, vol, cb, i, j, k) where {F} = begin
...
 
 add_z_vertex(m::MC{F}, vol, cb, i, j, k) where {F} = begin
     @inbounds begin
         u = cb[1] / (cb[1] - cb[5])
-        m.nrm[1] = (1 - u) * ∇x(vol, i, j, k, m.nx) + u * ∇x(vol, i, j, k + 1, m.nx)
-        m.nrm[2] = (1 - u) * ∇y(vol, i, j, k, m.ny) + u * ∇y(vol, i, j, k + 1, m.ny)
-        m.nrm[3] = (1 - u) * ∇z(vol, i, j, k, m.nz) + u * ∇z(vol, i, j, k + 1, m.nz)
+        m.nrm[1] = m.normal_sign * (1 - u) * ∇x(vol, i, j, k, m.nx) + u * ∇x(vol, i, j, k + 1, m.nx)
+        m.nrm[2] = m.normal_sign * (1 - u) * ∇y(vol, i, j, k, m.ny) + u * ∇y(vol, i, j, k + 1, m.ny)
+        m.nrm[3] = m.normal_sign * (1 - u) * ∇z(vol, i, j, k, m.nz) + u * ∇z(vol, i, j, k + 1, m.nz)
```
Comment (t-bltg) [inline src/MarchingCubes.jl:645]:
```suggestion
        m.nrm[1] = (1 - u) * ∇x(vol, i, j, k, m.nx) + u * ∇x(vol, i, j, k + 1, m.nx)
        m.nrm[2] = (1 - u) * ∇y(vol, i, j, k, m.ny) + u * ∇y(vol, i, j, k + 1, m.ny)
        m.nrm[3] = (1 - u) * ∇z(vol, i, j, k, m.nz) + u * ∇z(vol, i, j, k + 1, m.nz)
```

File: src/MarchingCubes.jl [hunk 9]
```diff
@@ -616,33 +620,33 @@ add_x_vertex(m::MC{F}, vol, cb, i, j, k) where {F} = begin
...
 
 add_y_vertex(m::MC{F}, vol, cb, i, j, k) where {F} = begin
     @inbounds begin
         u = cb[1] / (cb[1] - cb[4])
-        m.nrm[1] = (1 - u) * ∇x(vol, i, j, k, m.nx) + u * ∇x(vol, i, j + 1, k, m.nx)
-        m.nrm[2] = (1 - u) * ∇y(vol, i, j, k, m.ny) + u * ∇y(vol, i, j + 1, k, m.ny)
-        m.nrm[3] = (1 - u) * ∇z(vol, i, j, k, m.nz) + u * ∇z(vol, i, j + 1, k, m.nz)
+        m.nrm[1] = m.normal_sign * (1 - u) * ∇x(vol, i, j, k, m.nx) + u * ∇x(vol, i, j + 1, k, m.nx)
+        m.nrm[2] = m.normal_sign * (1 - u) * ∇y(vol, i, j, k, m.ny) + u * ∇y(vol, i, j + 1, k, m.ny)
+        m.nrm[3] = m.normal_sign * (1 - u) * ∇z(vol, i, j, k, m.nz) + u * ∇z(vol, i, j + 1, k, m.nz)
```
Comment (t-bltg) [inline src/MarchingCubes.jl:632]: Leftovers, to be removed
```suggestion
        m.nrm[1] = (1 - u) * ∇x(vol, i, j, k, m.nx) + u * ∇x(vol, i, j + 1, k, m.nx)
        m.nrm[2] = (1 - u) * ∇y(vol, i, j, k, m.ny) + u * ∇y(vol, i, j + 1, k, m.ny)
        m.nrm[3] = (1 - u) * ∇z(vol, i, j, k, m.nz) + u * ∇z(vol, i, j + 1, k, m.nz)
```

File: test/runtests.jl [hunk 10]
```diff
@@ -1,5 +1,6 @@
 using BenchmarkTools
 using GeometryBasics
+using LinearAlgebra
```
Comment (t-bltg) [inline test/runtests.jl:3]:
```suggestion

```

I'd prefer if we use `MarchingCubes.norm` instead of adding this dependency.

File: test/runtests.jl [hunk 11]
```diff
@@ -109,6 +110,21 @@ end
...
+    m1 = MC(dat)
+    march(m1)
+
+    m2 = MC(dat,normal_sign=-1)
+    march(m2)
+
+    @test all(norm.(m1.normals) .≈ norm.(m2.normals))
+
+    @test_throws ArgumentError MC(dat,normal_sign=2)
+    @test_throws ArgumentError MC(dat,normal_sign=-2)
```
Comment (t-bltg) [inline test/runtests.jl:125]: I'd prefer to be explicit about keyword arguments:
```suggestion
    @test_throws ArgumentError MC(dat; normal_sign=+2)
    @test_throws ArgumentError MC(dat; normal_sign=-2)
```

File: src/MarchingCubes.jl [hunk 12]
```diff
@@ -86,7 +88,9 @@ struct MC{F,I}
             Ref(x),
             Ref(y),
             Ref(z),
+            normal_sign,
         )
+        m.normal_sign ∈ (-1, +1) || throw(ArgumentError("`normal_sign` should be either -1 or +1"))
```
Comment (t-bltg) [inline src/MarchingCubes.jl:93]: This can be moved before the `m = ...` construct for early check.

```suggestion
        normal_sign ∈ (-1, +1) || throw(ArgumentError("`normal_sign` should be either -1 or +1")) 
```

**Review by t-bltg — CHANGES_REQUESTED** [review]

File: test/runtests.jl (hunk 10 above)
Comment (nickkeepfer) [reply test/runtests.jl:3]: Ahh, I didn't realise you could do that, great!

File: Project.toml [hunk 13]
```diff
@@ -1,11 +1,12 @@
 name = "MarchingCubes"
 uuid = "299715c1-40a9-479a-aaf9-4a633d36f717"
+license = "MIT"
 desc = "Efficient Implementation of Marching Cubes' Cases with Topological Guarantees"
 authors = ["T Bltg <tf.bltg@gmail.com>"]
-license = "MIT"
 version = "0.1.5"
 
 [deps]
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
```
Comment (t-bltg) [inline Project.toml:9]:
```suggestion
```

File: src/MarchingCubes.jl [hunk 14]
```diff
@@ -21,6 +21,7 @@ module MarchingCubes
 
 using SnoopPrecompile
 using StaticArrays
+using LinearAlgebra
```
Comment (t-bltg) [inline src/MarchingCubes.jl:24]:
```suggestion

```

File: test/runtests.jl [hunk 15]
```diff
@@ -109,6 +109,21 @@ end
...
+@testset "invert_normals" begin
+    dat = Float32[(x - 3)^2 + (y - 3)^2 + (z - 3)^2 for x ∈ 1:5, y ∈ 1:5, z ∈ 1:5]
+
+    m1 = MC(dat)
+    march(m1)
+
+    m2 = MC(dat,normal_sign=-1)
+    march(m2)
+
+    @test all(norm.(m1.normals) .≈ norm.(m2.normals))
```
Comment (t-bltg) [inline test/runtests.jl:121]:
```suggestion
    @test all(MarchingCubes.norm.(m1.normals) .≈ MarchingCubes.norm.(m2.normals))
```

**Comment by t-bltg** [comment]
Thanks for the work done here.

With CI passing, the only thing left to do is a visual check (posting an image here would be nice ;)).

**Review by t-bltg — APPROVED** [review]

**Review by t-bltg — APPROVED** [review]
Thanks for the work done here.

With CI passing, the only thing left to do is a visual check (posting an image here would be nice ;)).

**Comment by nickkeepfer** [comment]
> 

Thanks for your patience! 

Heres a figure to visually check, generated via:
```julia
fig = Figure(resolution=(800,400),fontsize = 22, figure_padding=(10,10,10,10))
ax1 = Axis3(fig[1,1],title="Normal")
ax2 = Axis3(fig[1,2],title="Inverted")
dat = Float32[(x - 3)^2 + (y - 3)^2 + (z - 3)^2 for x ∈ 1:5, y ∈ 1:5, z ∈ 1:5]
m1 = MC(dat,normal_sign=1)
m2 = MC(dat,normal_sign=-1)
march(m1,5.0)
march(m2,5.0)
msh1 = MarchingCubes.makemesh(GeometryBasics, m1)
msh2 = MarchingCubes.makemesh(GeometryBasics, m2)
mesh!(ax1,msh1)
mesh!(ax2,msh2)
```

Screenshot 2023-01-27 at 18 04 00

**Comment by t-bltg** [comment]
Great thanks.

I've checked that we have no regressions in [UnicodePlots](https://github.com/JuliaPlots/UnicodePlots.jl) (for [`isosurface`](https://github.com/JuliaPlots/UnicodePlots.jl#isosurface-plot)).

On paraview:
pv
